Refactor IDE docs & implement JetBrains private Beta docs

loujaybee opened the pull request:
## Description

This PR makes some fairly widespread changes to make way for the upcoming JetBrains IDE integrations. 

- [x] Create one page per integrated IDE
- [x] Introduce IDE section
- [x] Break down "JetBrains" page into one per IDE
- [x] Add pill to navigation to show if page is in Beta (**Note:** Consider separate PR for this)
- [x] Swap references to "in preview" for "beta"
- [x] Fix build
- [x] Note about Code With Me on JetBrains beta IDE's (e.g. data through "Code With Me" servers)
- [x] Add a link to a feedback issue on GitHub for beta IDE's
- [x] Update "coming soon" with information on when JetBrains will release their IDE's
- [x] Anything mentioned in [this internal thread](https://gitpod.slack.com/archives/C02CBKZEVS8/p1636460162111500)
- [x] Update "IDE" header for "editor" or similar (how do we incorporate vim here?)

### JetBrains Beta

- [x] What is Code With Me, what is the password?
- [x] JetBrains link will change on restart
- [x] Mention about licensing situation for JetBrains IDE's
- [x] Note about password `gitpod` implementation on JetBrains beta IDE's
- [x] How to enable JetBrains within Gitpod
- [x] A note about how the feature enables on new workspaces only (i.e restart if you change your IDE)
- [x] A note on licensing / current situation with the EAP

## Related Issue(s)

Fixes: gitpod-io/gitpod/issues/6524

## How to test

N/A

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

<a href="https://gitpod.io/#https://github.com/gitpod-io/website/pull/1225"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

loujaybee commented on src/routes/docs/menu.ts:
```diff
@@ -1,8 +1,14 @@
+  title: string,
+  path: string,
+  subMenu?: MenuEntry[],
+  status?: MenuStatus
```
Not a massive fan of having an extra argument like this, any suggestions on alternatives?

mikenikles replied:
It's ok, this is code we migrated from the old Gatsby site without much modifications.

A more Svelte-y way would likely be to use a store where we define the menu. Too much refactoring for this PR though.

loujaybee replied:
Thanks Mike, yeah I discussed with @nisarhassan12 and there wasn't an obvious easy answer for a better function signature right now 🙏

loujaybee commented on src/components/docs/menu-section.svelte:
```diff
@@ -39,6 +40,31 @@
+    margin-left: 12px;
+    background-color: var(--white);
+    color: var(--light-grey);
+    border: 1px solid var(--light-grey);
```
What's the take on spacing values: `px` vs `em` vs...? Any preference?

loujaybee commented on src/components/docs/menu-link.svelte:
```diff
@@ -15,4 +15,6 @@
 </style>
 
-<a class:active {href} sveltekit:prefetch {...$$props}><slot /></a>
+<a style="flex: 1;" class:active {href} sveltekit:prefetch {...$$props}
```
TODO: Remove inline style.

loujaybee commented on src/routes/docs/menu.ts:
```diff
@@ -30,16 +34,27 @@ export const MENU: MenuEntry[] = [
-    M("VS Code Desktop Support", "develop/vscode-desktop-support"),
+    M("Local Companion", "develop/local-companion", [], "beta" as MenuStatus),
+  ]),
+  M("IDE", "ide", [
```
Not a big fan of just "IDE" here however "Interactive Developer Environment" also seems very verbose.

loujaybee replied:
I slept on it... maybe "editor" or "editors" could also make sense here?

loujaybee commented on src/types/menu-entry.type.ts:
```diff
@@ -1,5 +1,11 @@
+export enum MenuStatus {
+  ComingSoon = "coming-soon",
```
"coming soon" feels long on the page, any shorter alternatives/suggestions?

![Screenshot 2021-11-08 at 18 43 53](https://user-images.githubusercontent.com/5528307/140799408-15c61b40-6e9a-4f4c-af3a-eaa7731768ed.png)

atduarte replied:
"Soon"?

loujaybee replied:
![image](https://user-images.githubusercontent.com/5528307/141103220-46b15fec-2c67-4cc1-b3e0-8687dc0f4e83.png)

loujaybee commented on src/routes/docs/menu.ts:
```diff
@@ -30,16 +34,27 @@ export const MENU: MenuEntry[] = [
+    M("RubyMine", "ide/rubymine", [], "coming-soon" as MenuStatus),
+    M("VS Code", "ide/vscode", [], "beta" as MenuStatus),
+    M("VS Code Extensions", "ide/vscode-extensions"),
+    M("WebStorm", "ide/webstorm", [], "coming-soon" as MenuStatus),
```
It's nice to have all the IDE's at the top-level of the docs, as this makes things quite obvious. However it means that applying a status label e.g. "coming soon" / "beta" is somewhat misleading, as we have the projector solutions for JetBrains. However, the projector solutions don't seem to be widely adopted, so maybe this is an okay compromise? 🤔  Any thoughts / ideas. Splitting into page for Browser/Desktop is a potential solution, but would really blow up the number of pages, which is already quite large.

loujaybee replied:
I might leave this as it is, following [this internal conversation](https://gitpod.slack.com/archives/C01KGM9ETU6/p1636622413102900). Feedback/suggestion is to have a separate "bring your own IDE" page, but we can potentially do this in a follow-up PR as not necessarily required to get this PR merged.

loujaybee commented on src/routes/docs/ide/vscode.md:
````diff
@@ -31,3 +33,16 @@ VS Code desktop support installs, manages and runs the [Local Companion](/docs/d
+   {
+     "editor.fontFamily": "Your custom font name"
+   }
+   ```
````
Removed this section from it's own page, however I'm not 100% sure why we really document this, seems like something we should just leave to the VSCode docs? 🤔

loujaybee commented on src/routes/docs/ide/clion.md:
```diff
@@ -0,0 +1,27 @@
+---
```
Since each IDE might potentially have a desktop and a browser section, it'd be nice to make that divide on the page quite obvious, maybe within each page we have three sections, intro (common docs), then browser, then desktop docs...

akosyakov commented:
It seems preview is failing: https://app.netlify.com/sites/gitpod-kumquat/deploys/61896a9de0a9d800083b035f

atduarte commented on src/routes/docs/ide/goland.md:
```diff
@@ -0,0 +1,27 @@
+
+## Desktop
+
+We currently have support for JetBrains GoLand (in private Beta). If you're interested to be part of the private Beta group, [leave a comment in this GitHub issue](https://github.com/gitpod-io/gitpod/issues/6342).
```
We are experiencing issues still, IIRC

loujaybee replied:
Fixed in: https://github.com/gitpod-io/gitpod/pull/6703

atduarte commented on src/routes/docs/ide/index.md:
```diff
@@ -0,0 +1,10 @@
+title: IDE
+---
+
+Gitpod isn't only a browser-based IDE, or 'VS Code in the browser'.
```
(Weakly-held) suggestion: "Gitpod is much more than a"

loujaybee replied:
Updated !

atduarte commented on src/routes/docs/ide/index.md:
```diff
@@ -0,0 +1,10 @@
+
+Gitpod isn't only a browser-based IDE, or 'VS Code in the browser'.
+
+In line with [our ambition to "remove all friction from the developer experience"](https://www.notion.so/gitpod/Values-Attributes-2ed4c2f93c84499b98e3b5389980992e), Gitpod will support many different IDE's and workflows.
```
Should it be IDEs' instead of IDE's ?

atduarte commented on src/routes/docs/ide/index.md:
```diff
@@ -0,0 +1,10 @@
+
+In line with [our ambition to "remove all friction from the developer experience"](https://www.notion.so/gitpod/Values-Attributes-2ed4c2f93c84499b98e3b5389980992e), Gitpod will support many different IDE's and workflows.
+
+Integrating every IDE natively isn't possible, as the IDE has to support remote development and/or run in a browser. However, Gitpod already supports many different type of IDE, both in your browser and via your desktop.
```
Suggestion: "isn't yet possible"

loujaybee replied:
Decided to drop the whole paragraph 🙏

atduarte commented on src/routes/docs/ide/intellij.md:
```diff
@@ -0,0 +1,27 @@
+
+## Desktop
+
+We currently have support for JetBrains IntelliJ IDEA (in private Beta). If you're interested to be part of the private Beta group, [leave a comment in this GitHub issue](https://github.com/gitpod-io/gitpod/issues/6342).
```
We should mention that IDEA Ultimate is/will be required.

atduarte commented on src/routes/docs/ide/vscode.md:
```diff
@@ -9,7 +9,9 @@ title: VS Code Desktop Support
-> **Please note:** VS Code Desktop Support is currently in preview If you experience any issues, or have feedback, please [get in touch](https://www.gitpod.io/support).
+## VSCode Desktop
+
+> **Please note:** VS Code Desktop Support is currently in beta If you experience any issues, or have feedback, please [get in touch](https://www.gitpod.io/support).
```
There's a missing dot between "in beta" and "If you".

loujaybee replied:
Added, good catch, thanks!

loujaybee commented:
> It seems preview is failing: https://app.netlify.com/sites/gitpod-kumquat/deploys/61896a9de0a9d800083b035f It would be good to fix it that we can have a look at the result.

Ah, I broke some links which the Netlify build caught (that's some nice #dx!). That's all cleaned up now. 

You can find the majority of the changes here: https://deploy-preview-1225--gitpod-kumquat.netlify.app/docs/ide

loujaybee commented on src/routes/docs/ide/jetbrains-private-beta.md:
```diff
@@ -0,0 +1,39 @@
+---
```
Page intentionally not linked to, will use as a reference for those in the private beta, but will eventually delete and incorporate into the main docs. I don't want to add this information into the main docs right now, as the current solution e.g. Code With Me etc might be confusing for other users who aren't in the Beta.

JohannesLandgraf commented:
Nice to see this coming together! 

- [ ] Align with @ChristinFrohne on design of the feature status (soon, beta etc)
- [ ] Make sure it works on mobile. @nisarhassan12 can help
- [ ] We decided last time to align with GitLab and call our 'beta' features 'preview' if I remember correctly. In any case we need to make a decision here and align (could be a very quick RFC that we can quickly) 
- [ ]  On VS Code we need to have both browser and desktop explained (the beta should be removed from the nav bar)
- [ ] Embed screencasts from Pauline where they make sense (VS Code Desktop, How extensions work etc)

pawlean commented:
Looking good! Just dropping here to apologise for giving you a conflict @loujaybee 😬 FWIW, I embedded the new VS Code Desktop Support and Extensions screencasts for you already! 🥳  I'm re-recording local companion app but should be ready soon.

atduarte commented on src/routes/docs/ide/goland.md:
```diff
@@ -0,0 +1,29 @@
+
+We currently have support for JetBrains GoLand (in private Beta). If you're interested to be part of the private Beta group, [leave a comment in this GitHub issue](https://github.com/gitpod-io/gitpod/issues/6342). If you have any issues with the IntelliJ GoLand Beta, you can [leave feedback in this GitHub issue](https://github.com/gitpod-io/gitpod/issues/6576).
+
+> **Important:** The current JetBrains private Beta works via [Code With Me](https://www.jetbrains.com/code-with-me/) to facilitate early Beta testing. The final JetBrains is expected not to require use of Code With Me. The Beta implementation means that anyone who might discover the workspace URL could potentially get access to your workspace. Therefore, you should only use the JetBrains Desktop ...
```
This sounds good 🎩  but felt a bit too long. May I suggest:
```
> **Warning:** Currently, we use [Code With Me](https://www.jetbrains.com/code-with-me/) links and a shared password to facilitate early Beta testing. This means that for now, anyone who might discover the workspace URL could potentially get full access to your workspace. This is temporary and will be resolved until the final release.
```

loujaybee replied:
Like it, thanks @atduarte ! Made a couple tweaks on top of your suggestion, but accepting the suggestion, looks good. Thanks for helping clean it up ! 

Committed version: 

```**Warning:** Currently, our JetBrains implementation uses [Code With Me](https://www.jetbrains.com/code-with-me/) links and a shared password to facilitate early Beta testing. This means that for now, anyone who might discover the workspace URL could potentially get full access to your workspace. This solution is temporary and will be resolved before the final release.```

ChristinFrohne commented:
I'd suggest using the same style as for the IDE preview on [gitpod.io](https://www.gitpod.io/) (Note: As GoLand and IntelliJ IDEA are now in beta, we'd need to add a screenshot of them as well. I can take care of that)

**Suggestion:**
To reduce noise, we could consider showing the "soon" label only on hover and active state. 

|default|hover & active state|
|-|-|
|<img width="262" alt="Bildschirmfoto 2021-11-11 um 13 27 22" src="https://user-images.githubusercontent.com/46345125/141298809-56cc952f-6200-4037-8cf0-8bb0d86e4978.png">| <img width="260" alt="Bildschirmfoto 2021-11-11 um 13 27 57" src="https://user-images.githubusercontent.com/46345125/141298841-0f531b72-713c-4b41-9a96-5de708217d1e.png">|

loujaybee commented:
Thanks for the feedback @ChristinFrohne !! Much appreciated! 🙏 

> To reduce noise, we could consider showing the "soon" label only on hover and active state.

I do have some apprehensions with regards to implementing as an active state: 

1. If we move to hover states, it then doesn't really solve my original intention of being able to "glance" over all the IDE's and get an idea of what's currently implemented, what's not, and what's upcoming 🤔  (in which case it might just be overkill / unnecessary for the PR and the right decision might just be to remove). 
2. How it might look on mobile (e.g. active state) I know hover states can be a bit fiddly on mobile as the "hover" isn't a native concept when you don't have a mouse 🤔 [[1](https://uxmovement.medium.com/the-hover-effect-for-mobile-buttons-ed9735fd5edc)]. But I wouldn't call myself an expert on usability best practices so would defer to someone like @gtsiolis or @nisarhassan12 to validate that/provide more feedback. 

I guess it's worth mentioning that this would be somewhat of a "temporary" thing until we get the JetBrains stuff through beta. But it is feasible that it'd hang around for at least a couple more months 🤔

loujaybee commented:
> On VS Code we need to have both browser and desktop explained (the beta should be removed from the nav bar) ([source](https://github.com/gitpod-io/website/pull/1225#issuecomment-966229487))

Added a separate page for Browser vs Desktop. Browser page could do with some added content, still...

![Screenshot 2021-11-17 at 21 27 02](https://user-images.githubusercontent.com/5528307/142285033-85c8afbd-739b-46d0-b64a-598b551e015a.png)

> I'd suggest using the same style as for the IDE preview on gitpod.io ([source](https://github.com/gitpod-io/website/pull/1225#issuecomment-966271470))

Updated 🙏 

![Screenshot 2021-11-17 at 21 29 19](https://user-images.githubusercontent.com/5528307/142285364-fa34aaf9-d3b9-421d-b9f6-2b44a21ef776.png)

loujaybee commented:
>  We decided last time to align with GitLab and call our 'beta' features 'preview' if I remember correctly. In any case we need to make a decision here and align (could be a very quick RFC that we can quickly approve)

A decision is pending here (See: [internal conversation](https://gitpod.slack.com/archives/C01KGM9DVRC/p1637185914139400))

loujaybee commented:
Things not yet solved, however these all don't seem like blockers for getting the PR merged. 

- [ ] "Could you send me the screenshots @loujaybee size: 1250 x 800)" - @ChristinFrohne 
- [ ] How do you install plugins?
- [ ] A note on how extensions work with the integration
- [ ] Some notes on the thin client / gateway, and implications
- [ ] Mention about [authentication with JetBrains](https://gitpod.slack.com/archives/C02CBKZEVS8/p1636977293136800?thread_ts=1636722452.117700&cid=C02CBKZEVS8)
- [ ] Update dashboard to incorporate new URL structure
- [ ] Validate all other pages are IDE agnostic
- [ ] Check casing of all IntelliJ products

loujaybee commented on _redirects:
```diff
@@ -22,12 +22,14 @@
 /docs/43_config_ports /docs/config-ports 301
 /docs/44_config_start_tasks /docs/config-start-tasks 301
 /docs/45_checkout_location /docs/checkout-location 301
+/docs/develop/vscode-desktop-support /docs/editors/vscode 301
```
This doesn't work for me on the local web app, will this only apply in prod? Or have I done something wrong? 

CC: @mikenikles

nisarhassan12 replied:
@loujaybee You have not done anything wrong the redirects are supposed to work on Prod and Netlify Deploy previews for the PRs.

loujaybee commented on src/components/docs/mobile-menu/sub-menu.svelte:
```diff
@@ -54,6 +54,30 @@
+  .status-beta {
+    border: none;
+    background-color: var(--brand-light);
+  }
```
This is duplicated in two places. 

I'm going to ask @filiptronicek if he's able to take a look and/or clean any of this up. 🙏

filiptronicek replied:
Sorry, @loujaybee, it looks like I am too late for this one, or is this still needed? 👀

loujaybee commented on src/routes/docs/editors/vscode.md:
````diff
@@ -33,3 +33,16 @@ VS Code desktop support installs, manages and runs the [Local Companion](/docs/d
+   {
+     "editor.fontFamily": "Your custom font name"
+   }
+   ```
````
Not new content, this is from: https://github.com/gitpod-io/website/pull/1225/files#diff-907f02a522964b63b9be89a733e60837acbd33dba8ad372097e14d572deec372

loujaybee commented on src/routes/docs/menu.ts:
```diff
@@ -30,16 +35,27 @@ export const MENU: MenuEntry[] = [
+    M("Local Companion", "develop/local-companion", [], "beta" as MenuStatus),
+  ]),
+  M("Editors", "editors", [
+    M("VS Code Browser", "editors/vscode-browser"),
```
Note from Anton about [VS Code Browser](https://gitpod.slack.com/archives/C01KGM9BH54/p1637221865122200?thread_ts=1637187168.121200&cid=C01KGM9BH54) wording.

nisarhassan12 commented on _redirects:
```diff
@@ -22,12 +22,16 @@
 /docs/47_Environment_Variables /docs/environment-variables 301
 /docs/50_ide /docs/ide 301
+/docs/ide /docs/editors 301
+/integrations/jetbrains /docs/editors 301
```
This is not gonna work. I think it should be: :)  
```suggestion
/docs/integrations/jetbrains /docs/editors 301
```

loujaybee replied:
Ah, good catch !

loujaybee commented:
@nisarhassan12, @mikenikles what are your thoughts on the styling changes in this PR? 

I've updated the branding of the changes to match the design that @ChristinFrohne mentioned [[1](https://github.com/gitpod-io/website/pull/1225#issuecomment-966271470)], however, it looks like the home page uses tailwind style classes, whereas I've added some classes manually here, what's the preferred approach? 

Do you have any guidance?

mikenikles commented on src/components/docs/mobile-menu/sub-menu.svelte:
```diff
@@ -107,6 +108,9 @@
         {#each currentSection?.subMenu as sub}
           <MenuItem href={sub.path} onClick={() => ($subMenuState = false)}>
             {sub.title}
+            {#if sub.status}
```
💡 If you move this into the `pill.svelte` component, you don't have to add this check wherever you use the `<Pill>` component.

I.e. make it a concern of the component as to whether or not it renders.

loujaybee replied:
That one also crossed my mind, but feels like it's balancing the too much functionality / abstraction in a component area 🤔  However, I've been away from the front-end world for some time, so I'd be more than happy to take others lead on this 🙏

mikenikles commented on src/routes/docs/editors/jetbrains-private-beta.md:
```diff
@@ -0,0 +1,39 @@
+---
+title: JetBrains Private Beta
```
I think we need a `section` here.

```suggestion
section: editors
title: JetBrains Private Beta
```

mikenikles reviewed (approved):
LGTM
